File: sdk/typescript/src/resource.ts
import type { MaybePromise } from "./api.ts";

export type DatastoreCapability =
  | "key_value"
  | "sql"
  | "blob_store";

export interface KVEntry {
  key: string;
  value: Uint8Array;
}

export interface KeyValueStore {
  get(key: string): MaybePromise<Uint8Array | null | undefined>;
  put(key: string, value: Uint8Array): MaybePromise<void>;
  delete(key: string): MaybePromise<void>;
  list(prefix?: string): MaybePromise<KVEntry[]>;
}

export interface SQLRows {
  columns: string[];
  rows: unknown[][];
}

export interface SQLExecResult {
  rowsAffected: number | bigint;
  lastInsertId: number | bigint;
}

export interface SQLMigration {
  version: number;
  description: string;
  sql: string;
}

export interface SQLStore {
  query(sql: string, params?: unknown[]): MaybePromise<SQLRows>;
  exec(sql: string, params?: unknown[]): MaybePromise<SQLExecResult>;
  migrate(migrations: SQLMigration[]): MaybePromise<void>;
}

export interface BlobEntry {
  key: string;
  size: number | bigint;
  contentType?: string;
  lastModified?: Date;
}

export interface BlobStore {
  get(key: string): MaybePromise<Uint8Array | null | undefined>;
  put(key: string, value: Uint8Array, contentType?: string): MaybePromise<void>;
  delete(key: string): MaybePromise<void>;
  list(prefix?: string): MaybePromise<BlobEntry[]>;
}

// ---------------------------------------------------------------------------
// Provider interfaces (namespace-aware, implemented by resource providers)
// ---------------------------------------------------------------------------

export interface DatastoreProvider {
  configure(name: string, config: Record<string, unknown>): MaybePromise<void>;
  capabilities(): DatastoreCapability[];
  healthCheck?(): MaybePromise<void>;
  close?(): MaybePromise<void>;
}

export interface KeyValueDatastoreProvider extends DatastoreProvider {
  kvGet(namespace: string, key: string): MaybePromise<Uint8Array | null | undefined>;
  kvPut(namespace: string, key: string, value: Uint8Array, ttlSeconds?: number): MaybePromise<void>;
  kvDelete(namespace: string, key: string): MaybePromise<void>;
  kvList(namespace: string, prefix?: string): MaybePromise<KVEntry[]>;
}

export interface SQLDatastoreProvider extends DatastoreProvider {
  sqlQuery(namespace: string, sql: string, params?: unknown[]): MaybePromise<SQLRows>;
  sqlExec(namespace: string, sql: string, params?: unknown[]): MaybePromise<SQLExecResult>;
  sqlMigrate(namespace: string, migrations: SQLMigration[]): MaybePromise<void>;
}

export interface BlobStoreDatastoreProvider extends DatastoreProvider {
  blobGet(namespace: string, key: string): MaybePromise<{ data: Uint8Array; contentType: string; metadata?: Record<string, string> } | null>;
  blobPut(namespace: string, key: string, value: Uint8Array, contentType?: string, metadata?: Record<string, string>): MaybePromise<void>;
  blobDelete(namespace: string, key: string): MaybePromise<void>;
  blobList(namespace: string, prefix?: string): MaybePromise<BlobEntry[]>;
}

File: sdk/rust/src/resource.rs
use std::collections::BTreeMap;
use std::time::SystemTime;

use serde::{Deserialize, Serialize};
use tonic::codegen::async_trait;

use crate::error::Result;

// ---------------------------------------------------------------------------
// Capability enum
// ---------------------------------------------------------------------------

#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DatastoreCapability {
    KeyValue,
    Sql,
    BlobStore,
}

// ---------------------------------------------------------------------------
// Key-value types and consumer trait
// ---------------------------------------------------------------------------

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct KvEntry {
    pub key: String,
    pub value: Vec<u8>,
}

#[async_trait]
pub trait KeyValueStore: Send + Sync {
    async fn get(&self, key: &str) -> Result<Option<Vec<u8>>>;
    async fn put(&self, key: &str, value: Vec<u8>) -> Result<()>;
    async fn put_with_ttl(&self, key: &str, value: Vec<u8>, ttl_seconds: i64) -> Result<()>;
    async fn delete(&self, key: &str) -> Result<bool>;
    async fn list(&self, prefix: &str, cursor: &str, limit: i32) -> Result<(Vec<KvEntry>, String)>;
}

// ---------------------------------------------------------------------------
// SQL types and consumer trait
// ---------------------------------------------------------------------------

#[derive(Clone, Debug, PartialEq)]
pub struct SqlRows {
    pub columns: Vec<String>,
    pub rows: Vec<Vec<serde_json::Value>>,
}

#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
pub struct SqlExecResult {
    pub rows_affected: i64,
    pub last_insert_id: i64,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct SqlMigration {
    pub version: i32,
    pub description: String,
    pub up_sql: String,
}

#[async_trait]
pub trait SqlStore: Send + Sync {
    async fn query(&self, query: &str, params: Vec<serde_json::Value>) -> Result<SqlRows>;
    async fn exec(&self, query: &str, params: Vec<serde_json::Value>) -> Result<SqlExecResult>;
    async fn migrate(&self, migrations: Vec<SqlMigration>) -> Result<()>;
}

// ---------------------------------------------------------------------------
// Blob types and consumer trait
// ---------------------------------------------------------------------------

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct BlobEntry {
    pub key: String,
    pub size: i64,
    pub content_type: String,
    pub last_modified: Option<SystemTime>,
}

#[async_trait]
pub trait BlobStore: Send + Sync {
    async fn get(&self, key: &str) -> Result<Option<(Vec<u8>, String)>>;
    async fn put(&self, key: &str, data: Vec<u8>, content_type: &str) -> Result<()>;
    async fn delete(&self, key: &str) -> Result<()>;
    async fn list(
        &self,
        prefix: &str,
        cursor: &str,
        limit: i32,
    ) -> Result<(Vec<BlobEntry>, String)>;
}

// ---------------------------------------------------------------------------
// Provider-facing traits (with namespace parameter)
// ---------------------------------------------------------------------------

#[async_trait]
pub trait DatastoreProvider: Send + Sync + 'static {
    async fn configure(
        &self,
        _name: &str,
        _config: serde_json::Map<String, serde_json::Value>,
    ) -> Result<()> {
        Ok(())
    }

    fn capabilities(&self) -> Vec<DatastoreCapability>;

    async fn health_check(&self) -> Result<()>;

    async fn close(&self) -> Result<()> {
        Ok(())
    }
}

#[async_trait]
pub trait KeyValueDatastoreProvider: DatastoreProvider {
    async fn kv_get(&self, namespace: &str, key: &str) -> Result<Option<Vec<u8>>>;
    async fn kv_put(
        &self,
        namespace: &str,
        key: &str,
        value: Vec<u8>,
        ttl_seconds: i64,
    ) -> Result<()>;
    async fn kv_delete(&self, namespace: &str, key: &str) -> Result<bool>;
    async fn kv_list(
        &self,
        namespace: &str,
        prefix: &str,
        cursor: &str,
        limit: i32,
    ) -> Result<(Vec<KvEntry>, String)>;
    async fn kv_migrate(&self, namespace: &str) -> Result<()>;
}

#[async_trait]
pub trait SqlDatastoreProvider: DatastoreProvider {
    async fn sql_query(
        &self,
        namespace: &str,
        query: &str,
        params: Vec<serde_json::Value>,
    ) -> Result<SqlRows>;
    async fn sql_exec(
        &self,
        namespace: &str,
        query: &str,
        params: Vec<serde_json::Value>,
    ) -> Result<SqlExecResult>;
    async fn sql_migrate(&self, namespace: &str, migrations: Vec<SqlMigration>) -> Result<()>;
}

#[async_trait]
pub trait BlobStoreDatastoreProvider: DatastoreProvider {
    async fn blob_get(
        &self,
        namespace: &str,
        key: &str,
    ) -> Result<Option<(Vec<u8>, String, BTreeMap<String, String>)>>;
    async fn blob_put(
        &self,
        namespace: &str,
        key: &str,
        data: Vec<u8>,
        content_type: &str,
        metadata: BTreeMap<String, String>,
    ) -> Result<()>;
    async fn blob_delete(&self, namespace: &str, key: &str) -> Result<()>;
    async fn blob_list(
        &self,
        namespace: &str,
        prefix: &str,
        cursor: &str,
        limit: i32,
    ) -> Result<(Vec<BlobEntry>, String)>;
}
